File: platforms/ios/Dezel/Dezel/Modules/View/WebView.swift
import WebKit

/**
 * @class WebView
 * @since 0.2.0
 * @hidden
 */
open class WebView: View, ContentWebViewDelegate {

	//--------------------------------------------------------------------------
	// MARK: Static Methods
	//--------------------------------------------------------------------------

	/**
	 * @method jsFunction_load
	 * @since 0.2.0
	 * @hidden
	 */
	public static func warmup() {
		WKWebView(frame: .zero, configuration: WKWebViewConfiguration()).loadHTMLString("<p>Warmup</p>", baseURL: nil)
	}

	//--------------------------------------------------------------------------
	// MARK: Properties
	//--------------------------------------------------------------------------

	/**
	 * @property view
	 * @since 0.2.0
	 * @hidden
	 */
	private var view: ContentWebView {
		return self.content as! ContentWebView
	}

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

	/**
	 * @inherited
	 * @method createContentView
	 * @since 0.5.0
	 */
	override open func createContentView() -> ContentWebView {
		return ContentWebView(frame: .zero, delegate: self)
	}

	/**
	 * @inherited
	 * @method dispose
	 * @since 0.6.0
	 */
	override open func dispose() {
		self.view.contentViewDelegate = nil
		self.view.stopLoading()
		super.dispose()
	}

	/**
	 * @inherited
	 * @method measure
	 * @since 0.5.0
	 */
	override open func measure(in bounds: CGSize, min: CGSize, max: CGSize) -> CGSize {
		return self.view.measure(in: bounds, min: min, max: max)
	}

	//--------------------------------------------------------------------------
	// MARK: WebView Delegate
	//--------------------------------------------------------------------------

	/**
	 * @method willLoad
	 * @since 0.2.0
	 * @hidden
	 */
	public func willLoad(webView: ContentWebView, url: URL) -> Bool {
		let result = self.context.createReturnValue()
		self.holder.callMethod("nativeBeforeLoad", arguments: [self.context.createString(url.absoluteString)], result: result)
		return result.boolean
	}

	/**
	 * @method didLoad
	 * @since 0.2.0
	 * @hidden
	 */
	public func didLoad(webView: ContentWebView) {
		self.holder.callMethod("nativeLoad")
	}

	/**
	 * @method didUpdateContentSize
	 * @since 0.2.0
	 * @hidden
	 */
	open func didUpdateContentSize(webView: ContentWebView, size: CGSize) {

		let contentWidth = Double(size.width)
		let contentHeight = Double(size.height)

		if (self.resolvedContentWidth != contentWidth) {
			if (self.layoutNode.wrapsContentWidth) {
				self.layoutNode.invalidate()
			}
		}

		if (self.resolvedContentHeight != contentHeight) {
			if (self.layoutNode.wrapsContentHeight) {
				self.layoutNode.invalidate()
			}
		}
	}

	//--------------------------------------------------------------------------
	// MARK: JS Methods
	//--------------------------------------------------------------------------

	/**
	 * @method jsFunction_load
	 * @since 0.2.0
	 * @hidden
	 */
	@objc open func jsFunction_load(callback: JavaScriptFunctionCallback) {
		self.view.load(url: callback.argument(0).string)
	}

	/**
	 * @method jsFunction_loadHTML
	 * @since 0.2.0
	 * @hidden
	 */
	@objc open func jsFunction_loadHTML(callback: JavaScriptFunctionCallback) {
		self.view.load(html: callback.argument(0).string)
	}

	/**
	 * @method jsFunction_reload
	 * @since 0.2.0
	 * @hidden
	 */
	@objc open func jsFunction_reload(callback: JavaScriptFunctionCallback) {
		self.view.reload()
	}

	/**
	 * @method jsFunction_stop
	 * @since 0.2.0
	 * @hidden
	 */
	@objc open func jsFunction_stop(callback: JavaScriptFunctionCallback) {
		self.view.stopLoading()
	}

	/**
	 * @method jsFunction_back
	 * @since 0.2.0
	 * @hidden
	 */
	@objc open func jsFunction_back(callback: JavaScriptFunctionCallback) {
		self.view.goBack()
	}

	/**
	 * @method jsFunction_forward
	 * @since 0.2.0
	 * @hidden
	 */
	@objc open func jsFunction_forward() {
		self.view.goForward()
	}
}

File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/modules/view/WebView.kt
package ca.logaritm.dezel.modules.view

import android.util.Size
import android.util.SizeF
import ca.logaritm.dezel.application.application
import ca.logaritm.dezel.core.JavaScriptContext
import ca.logaritm.dezel.core.JavaScriptFunctionCallback
import ca.logaritm.dezel.extension.ceiled
import ca.logaritm.dezel.view.ContentWebView
import ca.logaritm.dezel.view.ContentWebViewListener
import ca.logaritm.dezel.view.graphic.Convert

/**
 * @class WebView
 * @since 0.5.0
 * @hidden
 */
open class WebView(context: JavaScriptContext) : View(context), ContentWebViewListener {

	//--------------------------------------------------------------------------
	// Properties
	//--------------------------------------------------------------------------

	/**
	 * @property view
	 * @since 0.5.0
	 * @hidden
	 */
	private val view: ContentWebView
		get() = this.content as ContentWebView
	
	//--------------------------------------------------------------------------
	// Methods
	//--------------------------------------------------------------------------

	/**
	 * @inherited
	 * @method createContentView
	 * @since 0.5.0
	 */
	override fun createContentView(): ContentWebView {
		return ContentWebView(this.context.application, this)
	}

	/**
	 * @inherited
	 * @method destroy
	 * @since 0.5.0
	 */
	override fun dispose() {
		this.view.contentViewListener = null
		this.view.stopLoading()
		super.dispose()
	}

	/**
	 * @inherited
	 * @method measure
	 * @since 0.5.0
	 */
	override fun measure(bounds: SizeF, min: SizeF, max: SizeF): SizeF? {

		val size = this.view.measure(
			Convert.toPx(bounds),
			Convert.toPx(min),
			Convert.toPx(max)
		)

		return Convert.toDp(size).ceiled()
	}

	//--------------------------------------------------------------------------
	// WebView Listener
	//--------------------------------------------------------------------------

	/**
	 * @method willLoad
	 * @since 0.5.0
	 * @hidden
	 */
	override fun onBeforeLoad(webView: ContentWebView, url: String): Boolean {
		val result = this.context.createReturnValue()
		this.holder.callMethod("nativeBeforeLoad", arrayOf(this.context.createString(url)), result)
		return result.boolean
	}

	/**
	 * @method didLoad
	 * @since 0.5.0
	 * @hidden
	 */
	override fun onLoad(webView: ContentWebView) {
		this.holder.callMethod("nativeLoad")
	}

	/**
	 * @method didUpdateContentSize
	 * @since 0.5.0
	 * @hidden
	 */
	override fun onUpdateContentSize(webView: ContentWebView, size: Size) {

		val contentWidth = size.width.toDouble()
		val contentHeight = size.height.toDouble()

		if (this.resolvedContentWidth != contentWidth) {
			if (this.layoutNode.wrapsContentWidth) {
				this.layoutNode.invalidate()
			}
		}

		if (this.resolvedContentHeight != contentHeight) {
			if (this.layoutNode.wrapsContentHeight) {
				this.layoutNode.invalidate()
			}
		}
	}

	//--------------------------------------------------------------------------
	// JS Methods
	//--------------------------------------------------------------------------

	/**
	 * @method jsFunction_load
	 * @since 0.5.0
	 * @hidden
	 */
	@Suppress("unused")
 	open fun jsFunction_load(callback: JavaScriptFunctionCallback) {
		this.view.loadUrl(callback.argument(0).string)
	}

	/**
	 * @method jsFunction_loadHTML
	 * @since 0.5.0
	 * @hidden
	 */
	@Suppress("unused")
	open fun jsFunction_loadHTML(callback: JavaScriptFunctionCallback) {
		this.view.loadDataWithBaseURL("file:///android_asset/app", callback.argument(0).string, "text/html; charset=utf-8", "UTF-8", "")
	}

	/**
	 * @method jsFunction_reload
	 * @since 0.5.0
	 * @hidden
	 */
	@Suppress("unused")
	open fun jsFunction_reload(callback: JavaScriptFunctionCallback) {
		this.view.reload()
	}

	/**
	 * @method jsFunction_stop
	 * @since 0.5.0
	 * @hidden
	 */
	@Suppress("unused")
	open fun jsFunction_stop(callback: JavaScriptFunctionCallback) {
		this.view.stopLoading()
	}

	/**
	 * @method jsFunction_back
	 * @since 0.5.0
	 * @hidden
	 */
	@Suppress("unused")
	open fun jsFunction_back(callback: JavaScriptFunctionCallback) {
		this.view.goBack()
	}

	/**
	 * @method jsFunction_forward
	 * @since 0.5.0
	 * @hidden
	 */
	@Suppress("unused")
	open fun jsFunction_forward(callback: JavaScriptFunctionCallback) {
		this.view.goForward()
	}
}